plugins/package-managers/node/src/main/kotlin/yarn2/Yarn2DependencyHandler.kt

```diff
@@ -19,19 +19,80 @@
 
 package org.ossreviewtoolkit.plugins.packagemanagers.node.yarn2
 
+import java.io.File
+
 import org.ossreviewtoolkit.model.Identifier
 import org.ossreviewtoolkit.model.Issue
 import org.ossreviewtoolkit.model.Package
 import org.ossreviewtoolkit.model.PackageLinkage
 import org.ossreviewtoolkit.model.utils.DependencyHandler
+import org.ossreviewtoolkit.plugins.packagemanagers.node.NodePackageManagerType
+import org.ossreviewtoolkit.plugins.packagemanagers.node.PackageJson
+import org.ossreviewtoolkit.plugins.packagemanagers.node.moduleId
+import org.ossreviewtoolkit.plugins.packagemanagers.node.parsePackage
+import org.ossreviewtoolkit.plugins.packagemanagers.node.parsePackageJson
+
+internal class Yarn2DependencyHandler(private val yarn2: Yarn2) : DependencyHandler<PackageInfo> {
+    private val packageJsonForModuleId = mutableMapOf<String, PackageJson>()
+    private val moduleDirForModuleId = mutableMapOf<String, File>()
+    private val packageInfoForLocator = mutableMapOf<String, PackageInfo>()
+    private lateinit var workingDir: File
+
+    fun setContext(workingDir: File, moduleDirs: Set<File>, packageInfoForLocator: Map<String, PackageInfo>) {
+        this.workingDir = workingDir
+
+        this.packageInfoForLocator.apply {
+            clear()
+            putAll(packageInfoForLocator)
+        }
+
+        packageJsonForModuleId.clear()
+        moduleDirForModuleId.clear()
+
+        moduleDirs.forEach { moduleDir ->
+            val packageJson = parsePackageJson(moduleDir.resolve(NodePackageManagerType.DEFINITION_FILE))
+            packageJsonForModuleId[packageJson.moduleId] = packageJson
+            moduleDirForModuleId[packageJson.moduleId] = moduleDir
+        }
+    }
+
+    override fun identifierFor(dependency: PackageInfo): Identifier {
+        val moduleName = dependency.value.substringBeforeLast("@")
+        return Identifier(
+            type = if (dependency.isProject) yarn2.projectType else "NPM",
+            namespace = moduleName.substringBefore("/", ""),
+            name = moduleName.substringAfter("/"),
+            version = dependency.children.version
+        )
+    }
+
+    override fun dependenciesFor(dependency: PackageInfo): List<PackageInfo> =
+        dependency.children.dependencies.map { packageInfoForLocator.getValue(it.locator) }
+
+    override fun linkageFor(dependency: PackageInfo): PackageLinkage =
+        if (dependency.isProject) PackageLinkage.PROJECT_DYNAMIC else PackageLinkage.DYNAMIC
+
+    override fun createPackage(dependency: PackageInfo, issues: MutableCollection<Issue>): Package? {
+        val moduleId = dependency.moduleId
+        val packageJson = packageJsonForModuleId[moduleId]?.takeUnless { dependency.isProject } ?: return null
+
+        return parsePackage(
+            packageJson
+        ) { packageName: String ->
+            yarn2.getRemotePackageDetails(workingDir, setOf(packageName)).single()
+        }
```

Comment:
A bit more compact formatting:

        return parsePackage(packageJson) { packageName ->
            yarn2.getRemotePackageDetails(workingDir, setOf(packageName)).single()
        }

Reply:
I've addressed this here (as I was happy that the test are green and afraid of flaky red flags) https://github.com/oss-review-toolkit/ort/pull/10243